internal/uvm/network.go

```diff
@@ -24,8 +26,248 @@ var (
+	nsidInsideUVM := nsid
+	if vm.IsTemplate || vm.IsClone {
+		nsidInsideUVM = DEFAULT_CLONE_NETWORK_NAMESPACE_ID
+	}
+
+	// Query endpoints with actual nsid
+	endpoints, err := GetNamespaceEndpoints(ctx, nsid)
+	if err != nil {
+		return err
+	}
+
+	// Add the network namespace inside the UVM if it is not a clone. (Clones will
+	// inherit the namespace from template)
+	if !vm.IsClone {
+		// Get the namespace struct from the actual nsid.
+		hcnNamespace, err := hcn.GetNamespaceByID(nsid)
+		if err != nil {
+			return err
+		}
+
+		// All templates should have a special NSID so that it
+		// will be easier to debug. Override it here.
+		if vm.IsTemplate {
+			hcnNamespace.Id = nsidInsideUVM
+		}
+
+		if err = vm.AddNetNS(ctx, hcnNamespace); err != nil {
+			return err
+		}
+	}
+
+	// If adding a network endpoint to clones or a template override nsid associated
+	// with it.
+	if vm.IsClone || vm.IsTemplate {
+		// replace nsid for each endpoint
+		for _, ep := range endpoints {
+			ep.Namespace = &hns.Namespace{
+				ID: nsidInsideUVM,
+			}
+		}
+	}
+
+	if err = vm.AddEndpointsToNS(ctx, nsidInsideUVM, endpoints); err != nil {
+		// Best effort clean up the NS
+		if removeErr := vm.RemoveNetNS(ctx, nsidInsideUVM); removeErr != nil {
+			log.G(ctx).Warn(removeErr)
+		}
+		return err
+	}
+	return nil
+}
+
+// GetNamespaceEndpoints gets all endpoints in `netNS`
+func GetNamespaceEndpoints(ctx context.Context, netNS string) ([]*hns.HNSEndpoint, error) {
+	op := "uvm::GetNamespaceEndpoints"
+	l := log.G(ctx).WithField("netns-id", netNS)
+	l.Debug(op + " - Begin")
+	defer func() {
+		l.Debug(op + " - End")
+	}()
+
+	ids, err := hns.GetNamespaceEndpoints(netNS)
+	if err != nil {
+		return nil, err
+	}
+	var endpoints []*hns.HNSEndpoint
+	for _, id := range ids {
+		endpoint, err := hns.GetHNSEndpointByID(id)
+		if err != nil {
+			return nil, err
+		}
+		endpoints = append(endpoints, endpoint)
+	}
+	return endpoints, nil
+}
+
+// NCProxyEnabled returns if there is a network configuration client.
+func (uvm *UtilityVM) NCProxyEnabled() bool {
+	return uvm.ncProxyClient != nil
+}
+
+// NCProxyClient returns the network configuration proxy client.
+func (uvm *UtilityVM) NCProxyClient() ncproxyttrpc.NetworkConfigProxyService {
+	return uvm.ncProxyClient
+}
+
+// NetworkConfigType specifies the action to be performed during network configuration.
+// For example: setup or teardown
+type NetworkConfigType uint8
+
+const (
+	NetworkRequestSetup NetworkConfigType = iota
+	NetworkRequestTearDown
 )
 
+var ErrNoNetworkSetup = errors.New("no network setup present for UVM")
+
+// CreateAndAssignNetworkSetup creates and assigns a new NetworkSetup interface to the Utility VM.
+// This can be used to configure the networking (setup and teardown) of the vm.
+//
+// `addr` is an optional parameter
+func (uvm *UtilityVM) CreateAndAssignNetworkSetup(ctx context.Context, addr, containerID string) (err error) {
+	if uvm.NCProxyEnabled() {
```

Comment:
should we also enforce that `addr` is not empty here and fall back to local if it is for some reason?

Reply:
We can enforce addr isn't empty to fail earlier, but I don't think we should fallback